scheduler/strategy.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2015 The Serviced Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scheduler
+
+import (
+	"github.com/control-center/serviced/dao"
+	"github.com/control-center/serviced/datastore"
+	"github.com/control-center/serviced/domain/host"
+	"github.com/control-center/serviced/domain/service"
+	"github.com/control-center/serviced/domain/servicedefinition"
+	"github.com/control-center/serviced/facade"
+	"github.com/control-center/serviced/scheduler/strategy"
+	"github.com/zenoss/glog"
+)
+
+// Verify we implement all the interfaces
+var (
+	_ strategy.Host          = &StrategyHost{}
+	_ strategy.ServiceConfig = &StrategyRunningService{}
+	_ strategy.ServiceConfig = &StrategyService{}
+)
+
+type StrategyHost struct {
+	host     *host.Host
+	services []strategy.ServiceConfig
+}
+
+type StrategyRunningService struct {
+	svc *dao.RunningService
+}
+
+type StrategyService struct {
+	svc *service.Service
+}
+
+func StrategySelectHost(svc *service.Service, hosts []*host.Host, strat strategy.Strategy, facade *facade.Facade) (*host.Host, error) {
+
+	glog.V(2).Infof("Applying %s strategy for service %s", strat.Name(), svc.ID)
+
+	hostmap := map[string]*StrategyHost{}
+	hostids := []string{}
+
+	for _, h := range hosts {
+		// Create a StrategyHost for the host
+		hostmap[h.ID] = &StrategyHost{h, []strategy.ServiceConfig{}}
+		// Save off the hostid
+		hostids = append(hostids, h.ID)
+	}
+
+	// Look up all running services for the hosts
+	glog.V(2).Infof("Looking up instances for hosts: %+v", hostids)
+	svcs, err := facade.GetRunningServicesForHosts(datastore.Get(), hostids...)
+	if err != nil {
+		return nil, err
+	}
+	// Assign the services to the StrategyHosts
+	for _, s := range svcs {
+		h := hostmap[s.HostID]
```

Comment:
Seems like there might be a null-pointer risk here if, for whatever reason, s.HostID was not one of hosts[]